### ISSUE TITLE
MSTransferor Inconsistency with Rucio Rules for Recreated Rucio ules

### ISSUE THREAD

**Issue by hassan11196** [dmwm/WMCore#11763]
**Impact of the bug**
MSTransferor

There's an inconsistency between Rucio and MSTransferor. This happens when a transfer rule in Rucio is recreated for data that got stuck during the transfer process.

A specific request was stuck in the staging phase for over 150 days [1][2]. Upon investigation, it was discovered that one of its rules had become stuck and was then deleted. Felipe Gomez created a new rule for the transfer, and this time the data was successfully transferred. Rucio indicates that all 12 locks are functioning correctly [3], but MSTransferor is not showing 100% completion [3].

The question is: How can we make MSTransferor recognize and sync with the rules created outside of it in Rucio?

For now I have manually staged this workflow.

[1] Stuck request details: https://dmytro.web.cern.ch/dmytro/cmsprodmon/workflows.php?prep_id=task_B2G-RunIISummer20UL16MiniAODAPVv2-01421

[2] Jira Issue report: https://its.cern.ch/jira/browse/CMSPROD-71

[3] Rucio dataset: https://cms-rucio-webui.cern.ch/did?scope=cms&name=/ZprimeToZHToZlepHinc_narrow_M-4500_TuneCP5_13TeV-madgraph-pythia8/RunIISummer20UL16RECOAPV-106X_mcRun2_asymptotic_preVFP_v8-v2/AODSIM

[4] MSTransferor request info: https://cmsweb.cern.ch/ms-transferor/data/info?request=cmsunified_task_B2G-RunIISummer20UL16MiniAODAPVv2-01421__v1_T_230426_141813_9597

**Comment by amaltaro** [comment]
@hassan11196 Hassan, thanks for creating this issue.

Instead of having a system that would auto-discover rules that have been replaced **outside** WM, I am in favor of having an explicit update of the workflow transfer document such that MSMonitor can properly track all the relevant rules.

In other words, if someone deletes a Rucio rule responsible for bringing in input data for a production workflow, that person needs to also update the relevant document in the MSTransferor database. We still do not have an API for that though, but it's meant to be addressed with:
https://github.com/dmwm/WMCore/issues/11220

**Comment by hassan11196** [comment]
I agree this is a more cleaner approach, looking forward to its API implementation. 
Thank you for addressing this issue Alan.

**Comment by amaltaro** [comment]
Thinking twice, I was actually going to suggest to have both API and auto-discover mechanism, the problem is that with the rule gone, we have no way to identify what the DID name the rule was previously locking. So any auto-discover mechanism would also depend in the data discovery algorithm, which is just too complex and heavy to be continuously executed.

Hassan, if you agree, I would suggest to close this issue out.

**Comment by haozturk** [comment]
It's not ideal to ask the operator to do one more extra manual operation. My vote is on the implementation of auto-discover. It's not clear for me what you meant by data discovery algorithm and how it is different from auto-discover. MSMonitor already knows which rules it created. Doesn't it know which rule corresponds to which block? This doesn't have to be run frequently. It can be run once a day if the staging fraction is not growing in the last N iterations.

**Comment by vkuznet** [comment]
From design point of view it is better to provide an API which someone can use to update relevant document in MSTransferor database. Then, you need to decide on policy who will do this update. Here you have few options:
- implement auto mechanism which will do this with some interval
- add new API call to Rucio server which will make this call once it will get request to delete some rule
- open API to operators who may choose to add it to their workflows

Doing this way you properly allow handle all use-case and it will allow to address different use-case gradually, i.e. first, you'll create auto mechanism, but still can negotiate if such call will be done by operators, etc. I do not know if Rucio will allow to add such call in their codebase, but it is a possibility. Bottom line, you need to deal with service interactions via available APis.
